### ISSUE TITLE
Add support for GITHUB_TOKEN

### ISSUE THREAD

**Issue by jfechner-foxo** [hashicorp/setup-terraform#96]
I still think this should be an unresolved bug against the `setup-terraform` GitHub action.  While I agree that some of the comments in the original thread do provide a solution for the problem, they're ultimately janky at best, and don't really fit a best practices approach, given that you now have to create, essentially, an "admin" user, and proliferate SSH keys.

The GH action is supplied to make using terraform within GitHub actions easier, but doesn't automatically bootstrap the vendor supplied method for authenticating (`GITHUB_TOKEN`)?  Every other GH action makes use of this concept without forcing the problem onto its users.  I can understand the terraform project itself staying away from this behavior.  But not the GH action.

In any case, after running into this issue myself and hating the idea of scripting my own credential provider with bogus shell scripts or being forced to switch to SSH just for this, I stumbled upon this suggestion:
https://wahlnetwork.com/2020/08/11/using-private-git-repositories-as-terraform-modules/

Basically, you can inject your GITHUB_TOKEN into your git configuration with the following:
`git config --global url."https://oauth2:${{ secrets.GITHUB_TOKEN }}@github.com".insteadOf https://github.com`

Works like a charm.

_Originally posted by @doug-papenthien-by in https://github.com/hashicorp/setup-terraform/issues/33#issuecomment-714837963_

**Comment by TomTucka** [comment]
I totally think this would be a great addition to the action. Automatically bootstrapping the action to use `GITHUB_TOKEN` for authenticating would be brilliant. I think as @doug-papenthien-by pointed out, most other actions take advantage of this by default.

**Comment by mrchief** [comment]
For some weird reason, using 
```
git config --global url."https://oauth2:${TOKEN}@github.com".insteadOf https://github.com
``` 
works but using 
```
git config --global url."https://${USER}:${TOKEN}@github.com".insteadOf https://github.com
```
doesn't.

Also, `GITHUB_TOKEN` has limited scope so can only access current repo. If you're downloading another private repo, you need to create a PAT and use that instead. Further, the account (for which you create the PAT) needs to have at least read access to the other repo.

**Comment by collardmsc** [comment]
Being able to provide a github token would be great so we can use Github Apps for authentication

**Comment by davegallant** [comment]
I agree that being able to optionally pass in a token (much like [actions/checkout@](https://github.com/actions/checkout)) would be cleaner.

**Comment by davegallant** [comment]
I agree that being able to optionally pass in a token (much like [actions/checkout](https://github.com/actions/checkout)) would be cleaner.

**Comment by Satak** [comment]
Is there any way now to use GitHub App for authentication?

**Comment by dignajar** [comment]
Hello, I'm also interested in this feature. Thanks!

**Comment by dignajar** [comment]
I'm also interested in this feature, thanks!

**Comment by robbienohra** [comment]
Thanks for this, @jfechner-foxo! Worked like a charm — agreed it's quite odd that this isn't natively supported in the action. +1 from me to have this be natively supported.

**Comment by SteveDevOps** [comment]
I'm stumbling into this now as well as general confusion. Does this require the module pull to use https (not ssh) ? Could someone supply their yml/example(s) with this command in context?  My dilemma is I want to do local runs first for testing (with ssh source for module call).

**Comment by SteveDevOps** [comment]
okie.. solved all my headaches using https://github.com/webfactory/ssh-agent and set a deploy key on the module repo. Eessh.. what a day!

**Comment by atos-joshua-mendoza** [comment]
We could still really use this!
